Re-enable automatic substitutions in wxTextCtrl on OS X

vslavik opened the pull request:
This reverts ill-advised commits c07523734f2d93a8c9042d80fdd458954b0e1d41 and 8d42890df49564d6d81e413bcd4428bf6a8b10e6 that disabled native OS X behavior of substituting dashes and quotes with typographically correct characters _if the user has this feature enabled._

This was a bad idea for two reasons:
1. It made wx applications behave non-natively, and thus be worse, in a highly noticeable area.
2. It made it impossible for applications that want to behave correctly   to restore the native behavior, because once   `setAutomaticDashSubstitutionEnabled` or   `setAutomaticQuoteSubstitutionEnabled` is called (as `wxTextCtrl`   constructor does), it's no longer possible to obtain its original,   default setting.

It's not better to disable native functionality by default, it's better to be native and make customizations possible. `wxWindow` API exposes access to the native control and if an application desires to disable substitution behavior, it can easily do so from user code.

See also #186 and #241.

vslavik committed:
Re-enable automatic substitutions in wxTextCtrl on OS X

This reverts ill-advised commits
c07523734f2d93a8c9042d80fdd458954b0e1d41 and
8d42890df49564d6d81e413bcd4428bf6a8b10e6 that disabled native OS X
behavior of substituting dashes and quotes with typographically correct
characters if the user has this feature enabled.

This was a bad idea for two reasons:

1. It made wx applications behave non-natively, and thus be worse, in a
   highly noticeable area.

2. It made it impossible for applications that want to behave correctly
   to restore the native behavior, because once
   setAutomaticDashSubstitutionEnabled or
   setAutomaticQuoteSubstitutionEnabled is called (as wxTextCtrl
   constructor does), it's no longer possible to obtain its original,
   default setting.

It's not better to disable native functionality by default, it's better to
be native and make customizations possible. wxWindow API exposes access
to the native control and if an application desires to disable
substitution behavior, it can easily do so from user code.

See also #186 and #241.

vslavik commented:
(For reference, [here’s an example](https://github.com/vslavik/poedit/blob/69be3f6100ba2af555f9afcbb22c94bd4e8c2dea/src/text_control.cpp#L67) of entirely user-level code that (temporarily) disables the substitutions.)

vadz commented:
I don't really know what is the best thing to do here, so if you feel strongly that it's better to revert this, so be it. But I think it would be nice to at least mention this in the documentation, perhaps.

And it could also be useful to have some `wxTextEntry::UseForPlainText()` which would disable all this smartness, what do you think?

vslavik commented:
> if you feel strongly that it's better to revert this

I now do. When I first noticed this change, I shrugged it off as something easy enough to undo in user code, but I was wrong about it. The main issue I have with this is that it is impossible to restore default behavior (well, I suppose you could create a hidden dummy control to read its settings, but…). 

I definitely agree there *are* situations when this is undesirable (e.g. entering code), but it shouldn’t be the default, and such special cases should be explicitly enabled. Probably also worth nothing that a single-line text control has these off by default natively, so it only affects multiline/rich entry.

> But I think it would be nice to at least mention this in the documentation, perhaps.

I’m not sure. This is *normal* behavior on OS X (slash macOS, sigh), on par with things like long press for unusual characters entry etc., and has been for the past three years. Users are used to it, expect it, consider it normal… and know where to disable it if they don’t want it. I don’t think it’s wx docs’ place to document all native behavior. I suspect it’s mostly surprising to developers porting from other platforms.

> And it could also be useful to have some wxTextEntry::UseForPlainText() which would disable all this smartness, what do you think?

Nitpick: This is unrelated to *plain text* — I fully expect this to work even when the text being edited isn’t rich. That aside, I think being able to do it with a user-code Objective-C++ one-liner is good enough, but if you think it’s important enough to have dedicated `OSXEnableAutomaticXXX()` methods for it, I’ll add them. In that case I think it would be better to have the full 4-methods set instead of a single “disable all the substitutions” one.

vadz commented:
"Plain text" is not the best term, but I don't know how to say it better. "Dumb text"? Maybe "literal"?

And I think that a combined method which would be documented as disabling _all_ "smart substitutions", even those added by Apple in the future, would be most useful. It's indeed not that difficult to turn off the stuff the control does if you know about it, but who has time to track all changes in macOS (sic).

vslavik commented:
IMHO it doesn’t make sense to handle this as all-or-nothing. Even the PRs referenced above remarked that the user-customized `automaticTextReplacementEnabled` shouldn’t be messed with. It makes perfect sense to e.g. disable only smart quotes in a control expecting HTML markup (interferes with typing correct markup) and leave the rest enabled (doesn’t hurt anything). What would the *user’s* use case for disabling everything be? *Especially* if you don’t know what that may cover in the future?

Also FWIW, spelling and links detection are both *already* handled individually by wx.

Also[2], there’s value in using names like `OSXEnableAutomaticTextReplacement` that are similar to Cocoa’s methods that you’d find when looking for a way to disable the behavior — easier to find. See e.g. `OSXOnWillFinishLaunching` that is a thin wrapper around native API.

vadz commented:
Just to make it clear: I'm not at all against individual wrappers, but they're indeed trivial and we could live without them (living with them would be better though!).

My point is that if you have some text that really is supposed to be plain ASCII (for whatever reason but this is not as uncommon as we might wish it to be, use of Unicode can still be problematic in quite a few places), you really want to disable all smart quotes and smart everything else replacements.

vslavik committed:
Add API to control OS X wxTextCtrl’s smart behavior

Allow the user to customize smart quotes and dashes substutions on OS X
and also provide the OSXDisableAllSmartSubstitutions() method for
disabling them all at once.

vslavik commented:
Like this?

vadz commented:
Yes, looks good, thanks!

And when (not if, judging from their past form...) GTK+ follows Apple suit, we could then make the functions public and remove the "OSX" prefix...
